### PR TITLE
Channel health check improvement for cancelled requests

### PR DESCRIPTION
**Pull request by jeet1995**
# PR features

## Add channel health check gate when the channel sees request cancellations

In service-side node-freeze scenarios, connections / channels may not be severed thus causing the SDK to not detect unhealthiness on such channels, and with end-to-end timeout also configured for the request, such requests do not hit transit timeout either which could have otherwise helped deem the channel as unhealthy. 

In this PR, the SDK will track request cancellations and read delays at the channel level and if these properties exceed certain values, then the channel is deemed unhealthy. If such cancellations on a channel are a recurring symptom, the replica associated with this channel is treated as `Unhealthy` and is removed from the list of replicas to send a request to. This can help reduce the volume of timeouts seen due to end-to-end timeout configured.

## Replica validation optimization

Previously replica validation would be triggered for replicas in the `Unknown` health status despite the replica not being used by a container in the warm-up flow. This would create unnecessary connection attempts which could cause spikes in resource usage. Such replicas are excluded.
## Local testing done
- Verified that channel is being deemed unhealthy after 3 consecutive request cancellations on it and a read delay of more than 10s.
- Verified that channel being set as unhealthy is resulting in the replica's health status to be `Unhealthy`.
- Verified that replica being set as `Unhealthy` results in a background address refresh with the `forceAddress` flag as `true`.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 1]
```diff
@@ -5,14 +5,17 @@
...
 * Fixed request start time in the `CosmosDiagnostics` for individual request responses - See [PR 35705](https://github.com/Azure/azure-sdk-for-java/pull/35705)
 * Fixed an issue where `ConnectionStateListener` tracked staled `Uris` which fails to mark the current `Uris` unhealthy properly - See [PR 36067](https://github.com/Azure/azure-sdk-for-java/pull/36067)
 * Gone exceptions that are not idempotent should not be retried because it is not known if they succeeded for sure. The handling of the exception in this case is left to the user. Fixed retrying write operations when a gone exception occurs in bulk mode. - See [PR 35838](https://github.com/Azure/azure-sdk-for-java/pull/35838)
 * Fixed an issue to update the last unhealthy timestamp for an `Uri` instance only when transitioning to `Unhealthy` from a different health status -  See [36083](https://github.com/Azure/azure-sdk-for-java/pull/36083)
+* Improved the channel health check flow to deem a channel which sees cancellations as unhealthy. - See [PR 36225](https://github.com/Azure/azure-sdk-for-java/pull/36225)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:16]:
```suggestion
* Improved the channel health check flow to deem a channel unhealthy when seen consecutive cancellations. - See [PR 36225](https://github.com/Azure/azure-sdk-for-java/pull/36225)
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java [hunk 2]
```diff
@@ -979,16 +979,26 @@ private void validateReplicaAddresses(String collectionRid, AddressInformation[]
...
+
+                minConnectionsRequiredForEndpoint = this.connectionPolicy.getMinConnectionPoolSizePerEndpoint();
             }
 
             for (Uri addressToBeValidated : addressesNeedToValidation) {
 
+                // replica validation should be triggered for an address with Unknown health status only when
+                // the address is used by a container / collection which was part of the warm up flow
+                if (addressToBeValidated.getHealthStatus() == Uri.HealthStatus.Unknown) {
+                    if (!this.proactiveOpenConnectionsProcessor.isCollectionRidUnderOpenConnectionsFlow(collectionRid)) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java:991]: We can probably merge these two conditions, right?

if (addressToBeValidated == Unknown && !isCollectionUnderOpenConnectionsFlow()) {
continue;
}

Also, what if the replica is not in an Unknown status? We are still triggering the replica validation, right?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java [hunk 3]
```diff
@@ -397,6 +409,38 @@ private String idleConnectionValidation(Timestamps timestamps, Instant currentTi
         return errorMessage;
     }
 
+    private String isCancellationProneChannel(Timestamps timestamps, Instant currentTime, RntbdRequestManager requestManager, Channel channel) {
+        String errorMessage = StringUtils.EMPTY;
+
+        final Optional<RntbdContext> rntbdContext = requestManager.rntbdContext();
+
+        if (timestamps.cancellationCount() >= this.cancellationCountSinceLastReadThreshold) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java:417]: Worth adding this comment as what we have on the transitTimeout validation: 

// Transit timeout can be a normal symptom under high CPU load.
 // When request timeout due to high CPU, 
// close the existing the connection and re-establish a new one will not help the issue but rather make it worse, return fast

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java [hunk 4]
```diff
@@ -397,6 +409,38 @@ private String idleConnectionValidation(Timestamps timestamps, Instant currentTi
...
+
+            if (CpuMemoryMonitor.getCpuLoad().isCpuOverThreshold(this.timeoutDetectionDisableCPUThreshold)) {
+                return errorMessage;
+            }
+
+            final long readSuccessRecency = Duration.between(timestamps.lastChannelReadTime(), currentTime).toNanos();
+
+            if (readSuccessRecency > this.nonRespondingChannelReadDelayTimeLimitInNanos) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java:425]:
```suggestion
            if (readSuccessRecency >= this.nonRespondingChannelReadDelayTimeLimitInNanos) {
```

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @jeet1995  for the optimization

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java (hunk 2 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java:991]: `UnhealthyPending` is the other replica health status for which replica validation is triggered.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java (hunk 3 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java:417]: Fixed in next iteration.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java (hunk 4 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java]: Fixed in next iteration.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java (hunk 2 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java]: Merged `if` check in next iteration.

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 1 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md]: Fixed in next iteration.

**Pull request merged**